### PR TITLE
Introduce _pytest.timing as a way to control timing during internal tests

### PR DESCRIPTION
**Pull request by nicoddemus**
`_pytest.timing` is an indirection to 'time' functions, which pytest production
code should use instead of 'time' directly.

'mock_timing' is a new fixture which then mocks those functions, allowing us
to write time-reliable tests which run instantly and are not flaky.

This was triggered by recent flaky junitxml tests on Windows related to timing
issues:

https://github.com/pytest-dev/pytest/pull/7169/checks?check_run_id=698193136

```
================================== FAILURES ===================================
_______________________ TestPython.test_timing_function _______________________

self = <test_junitxml.TestPython object at 0x0000029912D92308>
testdir = <Testdir local('C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pytest-of-runneradmin\\pytest-0\\test_timing_function0')>
run_and_parse = <function run_and_parse.<locals>.run at 0x0000029912D88EE8>

    def test_timing_function(self, testdir, run_and_parse):
        testdir.makepyfile(
            """
            import time, pytest
            def setup_module():
                time.sleep(0.01)
            def teardown_module():
                time.sleep(0.01)
            def test_sleep():
                time.sleep(0.01)
        """
        )
        result, dom = run_and_parse()
        node = dom.find_first_by_tag("testsuite")
        tnode = node.find_first_by_tag("testcase")
        val = tnode["time"]
>       assert round(float(val), 2) >= 0.03
E       AssertionError: assert 0.02 >= 0.03
E        +  where 0.02 = round(0.024, 2)
E        +    where 0.024 = float('0.024')
```

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
CC @boxed for input from freezegun

**Comment by boxed** [comment]
Yea I would go full freezegun instead of this. Freezing all time functions solves this and much more. 

Here's an article I wrote for what we do at work: https://kodare.net/2018/09/12/flaky-tests-part-3-freeze-the-world.html

File: src/_pytest/timing.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+"""
+Indirection for time functions.
+
+We intentionally grab some "time" functions internally to avoid tests mocking "time" to affect
+pytest runtime information (issue #185).
+
+Fixture "mock_timinig" also interacts with this module for pytest's own tests.
+"""
+from time import perf_counter  # noqa: F401
```
Comment (bluetech) [inline src/_pytest/timing.py:9]: If you use `__all__`, you can avoid the noqa's.

**Comment by nicoddemus** [comment]
I love freezegun and use it all the time, it's an awesome library!

I'm not sure however it will work in this case in particular (but I will be glad if I'm wrong, as I would rather not reinvent stuff here).

There's a few requirements here thought that I think cannot be covered by `freezegun`:

1. We keep references to the original `time` functions we use on purpose (#185), so users mocking `time` or using `freezegun` on their own tests don't affect pytest (for example a user freezing time for a test should not change the total test session run time).

1. We are using `sleep()` inside the test functions to simulate running time. This can be circumvented by calling `freezegun` functions instead (such as `tick()`), but not sure this will then cover our cases well. 

Having said that, here's my attempt to convert one of the tests to use freeze gun:

```python
    def test_timing_function(self, testdir, run_and_parse, monkeypatch):
        from freezegun import freeze_time
        import sys

        with freeze_time("2020-05-23 12:00") as f:
            monkeypatch.setattr(sys, "pytest_frozen_time", f, raising=False)
            testdir.makepyfile(
                """
                import sys
                from datetime import timedelta
                def setup_module():
                    sys.pytest_frozen_time.tick(delta=timedelta(seconds=1))
                def teardown_module():
                    sys.pytest_frozen_time.tick(delta=timedelta(seconds=2))
                def test_sleep():
                    sys.pytest_frozen_time.tick(delta=timedelta(seconds=4))
            """
            )
            result, dom = run_and_parse()
            node = dom.find_first_by_tag("testsuite")
            tnode = node.find_first_by_tag("testcase")
            val = tnode["time"]
            assert float(val) == 7.0

```

This doesn't work (`assert 0.001 == 7.0`), and I suspect is that `freezegun` cannot (understandably) mock `_pytest.timing`. Of course having `_pytest.timing` shield `time.time` from mocking is the exact purpose why it was introduced.

Any suggestions @RonnyPfannschmidt @boxed? More than happy to apply those.

File: src/_pytest/timing.py (hunk 1 above)
Comment (nicoddemus) [reply src/_pytest/timing.py]: Ahh indeed, thanks!

**Comment by boxed** [comment]
Freezegun should give the real time we. Pytest asks for the time as relating to session timing. This has been a headache and a source of bugs though since it's hard to figure out where we are being called from. 

I don't really understand what this test code is doing so I don't understand where it's going wrong. I would expect the time as viewed from that tmp file python would be mocked as it shouldn't hit the whitelist.

**Comment by nicoddemus** [comment]
> This has been a headache and a source of bugs though since it's hard to figure out where we are being called from.

Indeed, and #185 is why we try to shield the time functions we use from `monkeypatching`/`freezegun`: users often patch (with `mock`, `monkeypatch` or `freezegun`) `time`-related functions in their own code, possibly in fixtures, and then are surprised that pytest is affected by it.

For example, `freezegun` might make time jump 10 days inside a test/fixture, which would then mess up the final test session time if we don't shield the time functions. We were already doing that in `runner.py`, in this PR I just extended that to other parts of the code.

So it is not really about using `freezegun` in some of our tests, it is about shielding timing functions from `monkeypatch`/`freezegun`.

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt gentle ping

**Comment by nicoddemus** [comment]
Backport: https://github.com/pytest-dev/pytest/pull/7306

**Pull request merged**